src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,310 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Runtime.InteropServices
+{
+    [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+    internal sealed partial class TypeMapLazyDictionary
+    {
+        private readonly LazyExternalTypeDictionary _externalTypeMap;
+        private readonly LazyProxyTypeDictionary _proxyTypeMap;
+
+        private ref struct CallbackContext
+        {
+            [RequiresUnreferencedCode("Lazy TypeMap isn't supported for Trimmer scenarios")]
+            public CallbackContext()
+            {
+                ExternalTypeMap = new LazyExternalTypeDictionary();
+                ProxyTypeMap = new LazyProxyTypeDictionary();
+            }
+
+            public LazyExternalTypeDictionary ExternalTypeMap;
+            public LazyProxyTypeDictionary ProxyTypeMap;
+        }
+
+        // See assemblynative.hpp for native version.
+        public unsafe struct ProcessAttributesCallbackArg
+        {
+            public void* Utf8String1;
+            public void* Utf8String2;
+            public void* Utf8String3;
+            public int StringLen1;
+            public int StringLen2;
+            public int StringLen3;
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "TypeMapLazyDictionary_ProcessAttributes")]
+        private static unsafe partial void ProcessAttributes(
+            QCallAssembly assembly,
+            QCallTypeHandle groupType,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newExternalTypeEntry,
+            delegate* unmanaged<CallbackContext*, ProcessAttributesCallbackArg*, void> newProxyTypeEntry,
+            CallbackContext* context);
+
+        private static TypeName CreateNewTypeName(string typeName, string assemblyName)
+        {
+            AssemblyNameInfo asmNameInfo = AssemblyNameInfo.Parse(assemblyName);
+            TypeName parsedType = new TypeName(typeName, asmNameInfo);
+            Debug.Assert(parsedType.AssemblyName != null);
+            return parsedType;
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NewExternalTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            string externalTypeName = new((sbyte*)arg->Utf8String1, 0, arg->StringLen1, Encoding.UTF8);
+            string targetType = new((sbyte*)arg->Utf8String2, 0, arg->StringLen2, Encoding.UTF8);
+
+            TypeName parsedTarget = TypeNameParser.Parse(targetType, throwOnError: true)!;
```

Comment:
`TypeName` is designed as a transient builder-like type. It carries a lot of waste for long-term storage. I am not sure whether it makes sense to use it as the key in the hashtable. I think it would make more sense to use more granular pieces (like just the type name) for the storage.

Reply:
`TypeName` isn't being used as a key, simply as a carrier of the needed key data. See the dictionary impl for what is used as the key.

Reply:
Ok, I have misread the code.